metpy/calc/tools.py

```diff
@@ -330,3 +332,142 @@ def log_interp(x, xp, fp, **kwargs):
         interpolated_vals = interpolated_vals * fp.units
 
     return interpolated_vals
+
+
+@exporter.export
+@check_units('[pressure]')
+def get_layer(p, *args, heights=None, bottom=None, depth=100 * units.hPa, interpolate=True):
+    r"""Return an atmospheric layer from upper air data with the requested bottom and depth.
+
+    This function will subset an upper air dataset to contain only the specified layer. The
+    bottom of the layer can be specified with a pressure or height above the surface
+    pressure. The bottom defaults to the surface pressure. The depth of the layer can be
+    specified in terms of pressure or height above the bottom of the layer. If the top and
+    bottom of the layer are not in the data, they are interpolated by default.
+
+    Parameters
+    ----------
+    p : array-like
+        Atmospheric pressure profile
+    datavar : array-like
+        Atmospheric variable measured at the given pressures
+    bottom : `pint.Quantity`
+        The bottom of the layer as a pressure or height above the surface pressure
+    depth : `pint.Quantity`
+        The thickness of the layer as a pressure or height above the bottom of the layer
+    interpolate : bool
+        Interpolate the top and bottom points if they are not in the given data
+
+    Returns
+    -------
+    `pint.Quantity, pint.Quantity`
+        The pressure and data variable of the layer
+
+    """
+    # Make sure pressure and datavar are the same length
+    for datavar in args:
+        if len(p) != len(datavar):
+            raise ValueError('Pressure and data variable must have the same length.')
+
+    # Bottom of layer calculations
+
+    # If no bottom is specified, use the first pressure value
+    if not bottom:
+        bottom_pressure = p[0]
+        if heights is not None:
+            bottom_height = heights[0]
+        else:
+            bottom_height = pressure_to_height_std(bottom_pressure)
+
+    # If the bottom pressure is specified
+    if bottom:
+        # in pressure units, assign it
+        if bottom.dimensionality == {'[length]': -1.0, '[mass]': 1.0, '[time]': -2.0}:
+            bottom_pressure = bottom
+            if heights is not None:
+                bottom_height = log_interp(bottom_pressure, p, heights)
+            else:
+                bottom_height = pressure_to_height_std(bottom_pressure)
+        # in height units
+        if bottom.dimensionality == {'[length]': 1.0}:
+            bottom_height = bottom
+            if heights is not None:
+                bottom_pressure = np.interp(np.array([bottom_height.m]), heights,
+                                            p)[0] * p.units
+            else:
+                bottom_pressure = height_to_pressure_std(bottom_height)
+
```

Comment:
Currently the `_get_bound_pressure_height` throws a value error if the bound has invalid units. Do we need to check here as well?